### PR TITLE
Add documentation for non tabular artifacts

### PR DESCRIPTION
**Pull request by HarisChoudhary**

### CODE REVIEW

File: artifacts.md [hunk 1]
```diff
@@ -6,7 +6,11 @@ description: Encapsulating data in Aqueduct
 
 _Artifacts_ are thin wrappers around data objects in Aqueduct. Wrapping data in an Artifact allows Aqueduct to track that artifact across workflow runs help you understand how your data is changing over time.&#x20;
 
-Aqueduct currently supports `TableArtifacts`, which are encapsulations of tabular or relational data, and [`ParameterArtifacts`](workflows/parameterizing-a-workflow.md), which wrap mutable json values. We plan to add support for other data types in the future.
+Aqueduct currently supports the following types of _Artifacts_:
+* `TableArtifacts`, which are encapsulations of tabular or relational data
+* `BoolArtifacts`, which are encapsulations of boolean data
+* `NumericArtifacts`, which are encapsulations of numeric data
+* `GenericArtifacts`, which are all other data type encapsulations that are not covered above. For supported data types, see [Artifact Data Types](artifacts/artifact-data-types.md)
```
Comment (vsreekanti) [inline artifacts.md:13]: This might be confusing. Rather than listing the class types, maybe we could just copy over the data types documentation here? What do you think?

File: operators/lazy-vs-eager-execution.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# Lazy vs Eager Executions
+
+Operators in python have two modes of execution supported by Aqueduct.
```
Comment (vsreekanti) [inline operators/lazy-vs-eager-execution.md:3]: Nit: capitalize Python.

File: operators/lazy-vs-eager-execution.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+* Lazy Execution: The function will not be run when invoked. The return type of the result Artifact will be `untyped`. When requesting the contents of this Artifact via the `.get()` command, Aqueduct will run the function and populate both the concrete type info and the value for the artifact
+* Eager Execution: The function will be run when invoked, the resultant artifact will already store the contents and the correct type information
+
+By default the mode of execution is `eager`. This can be configured via the `GlobalConfig` as shown below:
```
Comment (vsreekanti) [inline operators/lazy-vs-eager-execution.md:8]: Let's move this above the two bullet points.

File: operators/lazy-vs-eager-execution.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    return df
+
+# Function will not run on the Aqueduct backend yet so this call is non-blocking
+local_result = my_fn.local(pd.DataFrame())
+
+# Aqueduct will now execute my_fn to get the contents
+print(local_result.get())
+```
+
+Its important to keen in mind that [Metric & Checks](../metrics-and-checks.md) are similiar to operators and so follow the same execution modes as defined by the `GlobalConfig`.
```
Comment (vsreekanti) [inline operators/lazy-vs-eager-execution.md:28]: Can we also show an example with the `fn.lazy` syntax here?

**Review by vsreekanti — COMMENTED** [review]
Mostly looks good -- just a few changes requested inline!

File: operators/creating-a-python-operator.md [hunk 5]
```diff
@@ -12,17 +12,14 @@ def my_prediction_function(input: pd.DataFrame) -> pd.DataFrame:
...
-{% endhint %}
-
 When a Python function is annotated with `@op`, two things happen:
 
 1. Aqueduct packages up this code and infers any library dependencies that it might require, and prepares the code to run as a part of a workflow.&#x20;
 2. The annotated function will accept Aqueduct [Artifacts](../artifacts.md) as inputs and return Artifacts as outputs -- the return values can in turn by used by other operators.&#x20;
 
 {% hint style="info" %}
-Note that once an operator has been annotated with `@op`, it cannot be used as a "regular" Python function -- the inputs to the function must be Artifacts.
+Note that once an operator has been annotated with `@op`, it cannot be used as a "regular" Python function -- the inputs to the function can be either Artifacts or a non-Artifact.
```
Comment (kenxu95) [inline operators/creating-a-python-operator.md:21]: There can be multiple non-artifacts passed in, so maybe we don't use singular here? Also, saying something can "be either Artifacts or a non-Artifact" doesn't really give any new information. 

Can we sort of massage this into saying that the inputs to a function are expected to be an artifact, and if they aren't, we'll convert them into a parameter with the function argument name? It's also probably good to note that these functions always return artifacts, since that separates them from regular Python functions.

File: artifacts.md [hunk 6]
```diff
@@ -6,7 +6,19 @@ description: Encapsulating data in Aqueduct
 
 _Artifacts_ are thin wrappers around data objects in Aqueduct. Wrapping data in an Artifact allows Aqueduct to track that artifact across workflow runs help you understand how your data is changing over time.&#x20;
 
-Aqueduct currently supports `TableArtifacts`, which are encapsulations of tabular or relational data, and [`ParameterArtifacts`](workflows/parameterizing-a-workflow.md), which wrap mutable json values. We plan to add support for other data types in the future.
+All _Artifacts_ that are not of Type `GenericArtifact` are already aware of the underlying data types they hold. For example, the `BooleanArtifact` holds a `boolean`. A `GenericArtifact` holds a generic data type that is not encapsulated by the other explicit artifacts. A full list of Artifact data types within the system can be found below:
```
Comment (kenxu95) [inline artifacts.md:9]: hmm I think it flows much better if we don't talk about `GenericArtifact` and that stuff at all. To me, it reads like implementation detail that they should't care about. We should just have "Here is the full list of supported Artifact data types:" and then the list below.

And I think for the boolean and numeric entries, you can add a note on the side like
"boolean: This is also the type returned by Check operators.", etc

File: metrics-and-checks/checks-ensuring-correctness.md [hunk 7]
```diff
@@ -9,7 +9,7 @@ db = client.integration('aqueduct_demo')
 customers = db.sql('SELECT * FROM customers;')
 
 @check(severity=CheckSeverity.ERROR)
-def non_negative_workflows(customers: pd.DataFrame) -> boolean:
+def non_negative_workflows(customers: pd.DataFrame) -> bool:
```
Comment (kenxu95) [inline metrics-and-checks/checks-ensuring-correctness.md:12]: I feel like these bug fixes are more urgent, can we break them out in a separate PR and get those merged in faster?

File: operators/creating-a-python-operator.md [hunk 8]
```diff
@@ -12,17 +12,15 @@ def my_prediction_function(input: pd.DataFrame) -> pd.DataFrame:
...
-
 When a Python function is annotated with `@op`, two things happen:
 
 1. Aqueduct packages up this code and infers any library dependencies that it might require, and prepares the code to run as a part of a workflow.&#x20;
 2. The annotated function will accept Aqueduct [Artifacts](../artifacts.md) as inputs and return Artifacts as outputs -- the return values can in turn by used by other operators.&#x20;
 
 {% hint style="info" %}
-Note that once an operator has been annotated with `@op`, it cannot be used as a "regular" Python function -- the inputs to the function must be Artifacts.
+Note that once an operator has been annotated with `@op`, it cannot be used as a "regular" Python function.
+The inputs to this operator are expected to be Artifacts. In case you supply an input of another kind, we will implicitly convert the input into the appropiate Artifact representation.
```
Comment (kenxu95) [inline operators/creating-a-python-operator.md:22]: typo: "appropriate"

Could we slightly tweak this to: "If you supply a non-Artifact input, we will implicitly convert it into a workflow parameter for you, with the default value as the input."

File: operators/lazy-vs-eager-execution.md [hunk 9]
```diff
@@ -0,0 +1,27 @@
+# Lazy vs Eager Executions
+
+Operators in Python have two modes of execution supported by Aqueduct.
+By default the mode of execution is `eager`. This can be configured via the `GlobalConfig` as shown below.
```
Comment (kenxu95) [inline operators/lazy-vs-eager-execution.md:4]: But this is not the only way to configure it right? You can also run individual operators lazily with the `.lazy` annotate. Plus all the extract operators have a "lazy" argument.

File: operators/lazy-vs-eager-execution.md [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+# We can use the a boolean indicated by the key "lazy" to tell Aqueduct how we want execution to proceed
+aq.global_config({"lazy" : True})
+
+@aq.op
+def my_fn(df):
+    return df
+
+# Function will not run on the Aqueduct backend yet so this call is non-blocking
+local_result = my_fn.local(pd.DataFrame())
```
Comment (kenxu95) [inline operators/lazy-vs-eager-execution.md:21]: Did you mean `.lazy()` here? I don't get why we're showing `.local()` here.

**Review by kenxu95 — CHANGES_REQUESTED** [review]

**Comment by kenxu95** [comment]
Taking over this documentation task. I still need to update the example notebooks and paste them here, but everything else should be ready for feedback!

File: operators/lazy-vs-eager-execution.md [hunk 11]
```diff
@@ -0,0 +1,44 @@
+# Lazy vs Eager Execution
+
+The default mode of operator execution is `eager`. This means that whenever an operator is called, it will immediately run. This allows us to provide useful features like type enforcement.
```
Comment (vsreekanti) [inline operators/lazy-vs-eager-execution.md:3]: Let's add a small note saying something like "This is just like executing a regular Python function."

File: operators/lazy-vs-eager-execution.md [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+Alternatively, you may execute an operator in `lazy` mode. This means that we will only run that operator once the entire flow is published. This removes any pre-publish guardrails 
+that `eager` mode provides, but can greatly speed up flow creation process, since we no longer run the flow before publishing it.
```
Comment (vsreekanti) [inline operators/lazy-vs-eager-execution.md:6]: Slight edit: "For workflows processing a large amount of data, this can greatly sleep up the workflow creation process, since..."

File: parameters/creating-and-using-parameters.md [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+See this [**Workflow Parameters Tutorial**](../example-workflows/using-workflow-parameters.md) notebook for a full walkthrough of the API with concrete examples.
+
+### Creating Parameters
+
+Parameters are identified by their name, and *must* have default values. They can be created explicitly or implicitly.
+
+To create a parameter explicitly, use `client.create_param(<param_name>, default=<default_val>)`. These parameters can be fed into operators just like any other artifacts. 
+
+They can also be referenced within SQL queries using the double-bracket syntax! Eg. A query `select ... where col_name={{country_names}}` is completely valid, as long as a parameter named "country_names" exists.
+If "country_names" has current value `"'Argentina'"`, the query will implicitly expand into `select ... where col_name='Argentina'`. Note that SQL parameters can only be strings,
```
Comment (vsreekanti) [inline parameters/creating-and-using-parameters.md:12]: This sentence is slightly confusing. I might wonder: If it's a string, why do I have to insert quotes?

I think we want to clarify that the Python value is required to be a string. When doing interpolation, we don't do any type guards like adding quotes.

File: parameters/creating-and-using-parameters.md [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+and that if you need quotes around your parameter value, you will have to insert them yourself.
+
+We provide the following built-in SQL parameters:
+- `today`: expands to a string date in the format `'%Y-%m-%d'`. Eg. `SELECT * from hotel_reviews WHERE review_date={{ today }};`
+
+
+To create a parameter implicitly, feed a raw, non-artifact python value `v`into any operator. We will implicitly convert the input into a parameter for you, with the default value `v`.
```
Comment (vsreekanti) [inline parameters/creating-and-using-parameters.md:19]: Rather than "non-artifact", let's just say a regular Python object.

**Review by vsreekanti — APPROVED** [review]
Looks good, @kenxu95! I left a few small edits in line, but it should be good to go once those are in.

**Review by kenxu95 — APPROVED** [review]
